vllm/logger.py

```diff
@@ -1,61 +1,255 @@
-# Adapted from
-# https://github.com/skypilot-org/skypilot/blob/86dc0f6283a335e4aa37b3c10716f90999f48ab6/sky/sky_logging.py
 """Logging configuration for vLLM."""
+import inspect
+import json
 import logging
 import os
 import sys
+import time
+import traceback
+from pathlib import Path
+
+from loguru import logger
 
 VLLM_CONFIGURE_LOGGING = int(os.getenv("VLLM_CONFIGURE_LOGGING", "1"))
 
-_FORMAT = "%(levelname)s %(asctime)s %(filename)s:%(lineno)d] %(message)s"
-_DATE_FORMAT = "%m-%d %H:%M:%S"
 
+class InterceptHandler(logging.Handler):
+    """A custom logging handler for intercepting logs and redirecting them.
+
+    This handler is designed to integrate with loguru's `logger` library,
+    capturing log records generated by Python's standard logging module and
+    processing them using `logger`.
+
+    Attributes:
+        loglevel_mapping (dict): Mapping of logging level numbers to string
+        representations.
+    """
+    loglevel_mapping = {
+        50: "CRITICAL",
+        40: "ERROR",
+        30: "WARNING",
+        20: "INFO",
+        10: "DEBUG",
+        0: "NOTSET",
+    }
+
+    def emit(self, record: logging.LogRecord):
+        """Process the logging record.
+
+        Overrides the default emit method to handle log records. It binds the
+        logger to a specific name and adds contextual information.
+
+        Args:
+            record (logging.LogRecord): The log record to process.
+        """
+        try:
+            level = logger.level(record.levelname).name
+        except AttributeError:
+            level = self.loglevel_mapping[record.levelno]
+
+        frame, depth = inspect.currentframe(), 0
+        while frame and (depth == 0
+                         or frame.f_code.co_filename == logging.__file__):
+            frame = frame.f_back
+            depth += 1
+
+        logger.opt(depth=depth,
+                   exception=record.exc_info).log(level, record.getMessage())
```

Comment:
QQ: is `exception` arg necessary? Based on https://loguru.readthedocs.io/en/stable/api/logger.html#loguru._logger.Logger.opt, it says it is automatically resolved by `sys.exc_info()`?

```
exception ([bool](https://docs.python.org/3/library/functions.html#bool), [tuple](https://docs.python.org/3/library/stdtypes.html#tuple) or [Exception](https://docs.python.org/3/library/exceptions.html#Exception), optional) – If it does not evaluate as False, the passed exception is formatted and added to the log message. It could be an [Exception](https://docs.python.org/3/library/exceptions.html#Exception) object or a (type, value, traceback) tuple, otherwise the exception information is retrieved from [sys.exc_info()](https://docs.python.org/3/library/sys.html#sys.exc_info).
```